artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPTunneledCoreMessageReader.java

```diff
@@ -0,0 +1,160 @@
+ */
+public class AMQPTunneledCoreMessageReader implements MessageReader {
+
+   private final ProtonAbstractReceiver serverReceiver;
+
+   private boolean closed = true;
+   private DeliveryAnnotations deliveryAnnotations;
+
+   public AMQPTunneledCoreMessageReader(ProtonAbstractReceiver serverReceiver) {
+      this.serverReceiver = serverReceiver;
+   }
+
+   @Override
+   public DeliveryAnnotations getDeliveryAnnotations() {
+      return deliveryAnnotations;
+   }
+
+   @Override
+   public void close() {
+      closed = true;
+      deliveryAnnotations = null;
+   }
+
+   @Override
+   public MessageReader open() {
+      if (!closed) {
+         throw new IllegalStateException("Message reader must be properly closed before open call");
+      }
+
+      return this;
+   }
+
+   @Override
+   public Message readBytes(Delivery delivery) {
+      if (delivery.isPartial()) {
+         return null; // Only receive payload when complete
+      }
+
+      final AMQPSessionCallback sessionSPI = serverReceiver.getSessionContext().getSessionSPI();
+      final Receiver receiver = ((Receiver) delivery.getLink());
+      final ReadableBuffer recievedBuffer = receiver.recv();
+
+      if (recievedBuffer.remaining() == 0) {
+         throw new IllegalArgumentException("Received empty delivery when expecting a core message encoding");
+      }
+
+      final DecoderImpl decoder = TLSEncode.getDecoder();
+
+      decoder.setBuffer(recievedBuffer);
+
+      Data payloadData = null;
+
+      try {
+         while (recievedBuffer.hasRemaining()) {
+            final TypeConstructor<?> constructor = decoder.readConstructor();
+
+            if (Header.class.equals(constructor.getTypeClass())) {
+               constructor.skipValue(); // Ignore for forward compatibility
+            } else if (DeliveryAnnotations.class.equals(constructor.getTypeClass())) {
+               deliveryAnnotations = (DeliveryAnnotations) constructor.readValue();
+            } else if (MessageAnnotations.class.equals(constructor.getTypeClass())) {
+               constructor.skipValue(); // Ignore for forward compatibility
+            } else if (Properties.class.equals(constructor.getTypeClass())) {
+               constructor.skipValue(); // Ignore for forward compatibility
+            } else if (ApplicationProperties.class.equals(constructor.getTypeClass())) {
+               constructor.skipValue(); // Ignore for forward compatibility
+            } else if (Data.class.equals(constructor.getTypeClass())) {
+               if (payloadData != null) {
+                  throw new IllegalArgumentException("Received an unexpected additional Data section in core tunnled AMQP message");
+               }
+
+               payloadData = (Data) constructor.readValue();
+            } else if (AmqpValue.class.equals(constructor.getTypeClass())) {
+               throw new IllegalArgumentException("Received an AmqpValue payload in core tunnled AMQP message");
```

Comment:
typo: tunnled  (multiple places in this and other classes, suggest a find+replace on entire diff)